Readme.md
# Polygonscan API

## Development of a NEXTGEN Version has started - please stand by

A way to access the [polygonscan.com api](https://polygonscan.com/apis) using promises. Fetch a diverse set of information about the blockchain.

Mainnet

```javascript
var api = require("polygonscan-api").init("YourApiKey");
var balance = api.account.balance("0x0000000000000000000000000000000000001010");
balance.then(function (balanceData) {
  console.log(balanceData);
});
```

## For testnet usage

Supported:

Latest

```javascript
// apikey, network, timeout
var api = require('polygonscan-api').init('YourApiKey','mumbai'. '3000');
```

## Install

```bash
npm install polygonscan-api --save
```

## API Documentation

## Development workflow

- npm test - runs tests
  - npm run posttest - starts the linter
- npm run lint - preconfigured linter
- npm run docs - generates the apidocs
- npm run bundle - builds a new bundle
- npm run preversion - Steps before we create a new Tag
  - lint
  - changelog
- npm run pages - pushes generated apidocs to the server
- postversion - after generating a new version, push the tag to the server
- npm run changelog - generates a changelog and pushes it


docs/CHANGELOG.md
# 2021-08-24

- Initial Releases


test/eth-test.js
'use strict';
const assert = require('chai').assert;
const init = require('../.').init;
describe('proxy', function() {
  it('eth.getminedblocks', function(done){
    var api = init();
    var txlist = api.account.getminedblocks('0x46a3a41bd932244dd08186e4c19f1a7e48cbcdf4');
    txlist.then(function(res){
      assert.ok(res);
      done();
    });
  });
});


test/balance-test.js
'use strict';
const assert = require('chai').assert;
const init = require('../.').init;
describe('balance', function() {
  it('returns a promise', function( ){
    var api = init();
    var balance = api.account.balance('0x0000000000000000000000000000000000001010');
    assert.ok(balance.then);
  });

  it('executes the promise', function(done){
    var api = init();
    var balance = api.account.balance('0x0000000000000000000000000000000000001010');
    balance.then(function(){
      done();
    });
  });

  it('has data', function(done){
    var api = init();
    var balance = api.account.balance('0x0000000000000000000000000000000000001010');
    balance.then(function(res){
      assert.ok(res);
      done();
    });
  });

});


test/methods-test.js
'use strict';
const assert = require('chai').assert;
const init = require('../.').init;
describe('api', function() {

  describe('account', function() {
    it('getminedblocks', function(done){
      var api = init();
      var txlist = api.account.getminedblocks('0x46a3a41bd932244dd08186e4c19f1a7e48cbcdf4');
      txlist.then(function(res){
        assert.ok(res);
        done();
      });
    });
    /**
     * @deprecated by Polygonscan
     * https://polygonscan.com/apis#tokens
     */
    // it('tokenbalance by name', function (done) {
    //   var api = init();
    //
    //   var supply = api.account.tokenbalance(
    //     '0x0000000000000000000000000000000000001010',
    //     'DGD'
    //   );
    //   supply.then(function (res) {
    //     assert.ok(res.result);
    //     done();
    //   });
    // });
    it('tokenbalance by address', function(done) {
      var api = init();
      var supply = api.account.tokenbalance('0x0000000000000000000000000000000000001010', false, '0x8f3cf7ad23cd3cadbd9735aff958023239c6a063');
      supply.then(function(res){
        assert.ok(res);
        done();
      });
    });
    it('txlist', function(done){
      var api = init();
      var txlist = api.account.txlist('0x0000000000000000000000000000000000001010');
      txlist.then(function(res){
        assert.ok(res);
        done();
      });
    });
    it('txlistinternal by hash', function(done){
      var api = init();
      var txlist = api.account.txlistinternal('0x40eb908387324f2b575b4879cd9d7188f69c8fc9d87c901b9e2daaea4b442170');
      txlist.then(function(res){
        assert.ok(res);
        done();
      }).catch(done);
    });
    it('txlistinternal by address', function(done){
      var api = init();
      var txlist = api.account.txlistinternal(null, '0x0000000000000000000000000000000000001010');
      txlist.then(function(res) {
        assert.ok(res);
        done();
      });
    });
    it('balance', function(done){
      var api = init();
      var balance = api.account.balance('0x0000000000000000000000000000000000001010');
      balance.then(function(res){
        assert.ok(res);
        done();
      });
    });
    it('balance multi', function(done){
      var api = init();
      var balance = api.account.balance(['0x0000000000000000000000000000000000001010']);
      balance.then(function(res){
        assert.ok(res);
        done();
      });
    });
    it('tokentx', function(done){
      var api = init();
      var txlist = api.account.tokentx(
        '0x0000000000000000000000000000000000001010',
        '0x2791bca1f2de4661ed88a30c99a7a9449aa84174',
        1, 'latest', 'asc'
      );
      txlist.then(function(res){
        assert.ok(res);
        done();
      });
    });
  });
  describe('stats', function(){
    it('ethsupply', function(done){
      var api = init();
      var supply = api.stats.ethsupply();
      supply.then(function(res){
        assert.ok(res);
        done();
      });
    });

    it('tokensupply by tokenname', function(done){
      var api = init();
      var supply = api.stats.tokensupply('MKR');
      supply.then(function(res){
        assert.ok(res);
        done();
      });
    });
    it('tokensupply by address', function(done){
      var api = init();
      var supply = api.stats.tokensupply(null, '0x0000000000000000000000000000000000001010');
      supply.then(function(res){
        assert.ok(res);
        done();
      }).catch(done);
    });

    it('ethprice', function(done){
      var api = init();
      var price = api.stats.ethprice();
      price.then(function(res){
        assert.ok(res);
        done();
      }).catch(done);
    });
  });


  it('block.getblockreward', function(done){
    var api = init();
    var blockreward = api.block.getblockreward();
    blockreward.then(function(res){
      assert.ok(res);
      done();
    });
  });

  it('transaction.getstatus', function(done){
    var api = init();
    var status = api.transaction.getstatus('0x591ca00bf6b404e24e21732023abc0416673beff9586f37e61fb0f07ca560940');
    status.then(function(res){
      assert.ok(res);
      done();
    });
  });
  // test for bug #31
  // see https://github.com/sebs/etherscan-api/issues/31
  it('contract.getabi for a contract that is not verified by etherscan: error', function(done){
    var api = init();
    var abi = api.contract.getabi('0x2791bca1f2de4661ed88a30c99a7a9449aa84174');
    abi.then(function(){
      assert.false(true, 'should not be a success');
    }).catch(err=> {
      assert.equal(err, 'Contract source code not verified');
      done();
    });
  });



  describe('proxy', function() {
    it('proxy.eth_blockNumber', function(done){
      var api = init();
      var res = api.proxy.eth_blockNumber();
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });


    it('proxy.eth_getBlockByNumber', function(done){
      var api = init();
      var res = api.proxy.eth_getBlockByNumber('0x10d4f');
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });


    it('proxy.eth_getUncleByBlockNumberAndIndex', function(done){
      var api = init();
      var res = api.proxy.eth_getUncleByBlockNumberAndIndex('0x210A9B', '0x0');
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });

    it('proxy.eth_getBlockTransactionCountByNumber', function(done){
      var api = init();
      var res = api.proxy.eth_getBlockTransactionCountByNumber('0x10FB78');
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });

    it('proxy.eth_getTransactionByHash', function(done){
      var api = init();
      var res = api.proxy.eth_getTransactionByHash('0x1e2910a262b1008d0616a0beb24c1a491d78771baa54a33e66065e03b1f46bc1');
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });


    it('proxy.eth_getTransactionByBlockNumberAndIndex', function(done){
      var api = init();
      var res = api.proxy.eth_getTransactionByBlockNumberAndIndex('0x10d4f', '0x0');
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });

    it('proxy.eth_getTransactionCount', function(done){
      var api = init();
      var res = api.proxy.eth_getTransactionCount('0x2910543af39aba0cd09dbb2d50200b3e800a63d2');
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });

    xit('proxy.eth_sendRawTransaction', function(done){
      var api = init();
      var res = api.proxy.eth_sendRawTransaction('0xf904808000831cfde080');
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });

    it('proxy.eth_getTransactionReceipt', function(done){
      var api = init();
      var res = api.proxy.eth_getTransactionReceipt('0x1e2910a262b1008d0616a0beb24c1a491d78771baa54a33e66065e03b1f46bc1');
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });

    it('proxy.eth_call', function(done){
      var api = init();
      var res = api.proxy.eth_call(
        '0xAEEF46DB4855E25702F8237E8f403FddcaF931C0',
        '0x70a08231000000000000000000000000e16359506c028e51f16be38986ec5746251e9724',
        'latest'
      );
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });

    it('proxy.eth_getCode', function(done){
      var api = init();
      var res = api.proxy.eth_getCode('0xf75e354c5edc8efed9b59ee9f67a80845ade7d0c', 'latest');
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });

    it('proxy.eth_getStorageAt', function(done){
      var api = init();
      var res = api.proxy.eth_getStorageAt('0x6e03d9cce9d60f3e9f2597e13cd4c54c55330cfd', '0x0', 'latest');
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });

    it('proxy.eth_gasPrice', function(done){
      var api = init();
      var res = api.proxy.eth_gasPrice();
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });

    xit('proxy.eth_estimateGas', function(done){
      var api = init();
      var res = api.proxy.eth_estimateGas(
        '0xf0160428a8552ac9bb7e050d90eeade4ddd52843',
        '0xff22',
        '0x051da038cc',
        '0xffffff'
      );
      res.then(function(res){
        assert.ok(res);
        done();
      });
    });
  });
});


lib/init.js
"use strict";
const proxy = require('./proxy');
const stats = require('./stats');
const transaction = require('./transaction');
const contract = require('./contract');
const account = require('./account');
/**
 * @module polygonscan/api
 */

/**
 * @param {string} apiKey - (optional) Your Polygonscan APIkey
 * @param {string} chain - (optional) Testnet chain keys [ropsten, mumbai, kovan]
 * @param {number} timeout - (optional) Timeout in milliseconds for requests, default 10000
 */
module.exports = function(apiKey, chain, timeout) {

  if (!apiKey) {
    apiKey = "YourApiKey";
  }


  if (!timeout) {
    timeout = 10000;
  }

  var getRequest = require('./get-request')(chain, timeout);

  /** @lends module:polygonscan/api */
  return {
    /**
     * @namespace
     */
    proxy: proxy(getRequest, apiKey),
    /**
     * @namespace
     */
    stats: stats(getRequest, apiKey),
    /**
     * @namespace
     */
    transaction: transaction(getRequest, apiKey),
    /**
     * @namespace
     */
    contract: contract(getRequest, apiKey),
    /**
     * @namespace
     */
    account: account(getRequest, apiKey)
  };
};


lib/stats.js
const querystring = require('querystring');
module.exports = function(getRequest, apiKey) {
  return {
    /**
     * Returns the supply of Tokens
     * @param {string} tokenname - Name of the Token
     * @param {string} contractaddress - Address from token contract
     * @example
     * var supply = api.stats.tokensupply(null, '0x0000000000000000000000000000000000001010');
     * @returns {Promise.<object>}
     */
    tokensupply(tokenname, contractaddress) {
      const module = 'stats';
      const action = 'tokensupply';

      let params = {
        module, action, apiKey
      };

      if (tokenname) {
        params.tokenname = tokenname;
      }

      if (contractaddress) {
        params.contractaddress = contractaddress;
      }

      var query = querystring.stringify(params);
      return getRequest(query);
    },
  };
};


lib/get-request.js

const axios = require('axios');
/**
 * @param {string} chain
 * @returns {string}
 */
function pickChainUrl(chain) {
  if (!chain || !TESTNET_API_URL_MAP[chain]) {
    return MAIN_API_URL;
  }

  return TESTNET_API_URL_MAP[chain];
}


const MAIN_API_URL = 'https://api.polygonscan.com';
const TESTNET_API_URL_MAP = {
  mumbai: 'https://api-testnet.polygonscan.com',
};

module.exports = function(chain, timeout) {
  var client = axios.create({
    baseURL: pickChainUrl(chain),
    timeout: timeout
  });

  /**
   * @param query
   * @returns {Promise<any>}
   */
  function getRequest(query) {
    return new Promise(function(resolve, reject) {
      client.get('/api?' + query).then(function(response) {
        var data = response.data;
        console.log("echo query: ", '/api?' + query);


        if (data.status && data.status != 1) {
          let returnMessage = data.message ||'NOTOK';
          if (data.result && typeof data.result === 'string') {
            returnMessage = data.result;
          } else if (data.message && typeof data.message === 'string') {
            returnMessage = data.message;
          }

          return reject(returnMessage);
        }

        if (data.error) {
          var message = data.error;

          if(typeof data.error === 'object' && data.error.message){
            message = data.error.message;
          }

          return reject(new Error(message));
        }

        resolve(data);
      }).catch(function(error) {
        return reject(new Error(error));
      });
    });
  }

  return getRequest;
};


lib/account.js
const querystring = require('querystring');
module.exports = function(getRequest, apiKey) {
  return {
    /**
     * Returns the amount of Tokens a specific account owns.
     * @param {string} address - Contract address
     * @param {string} tokenname - Name of the token
     * @param {string} contractaddress - Contract address
     * @example
     *     var supply = api.account.tokenbalance(
     *       '0x0000000000000000000000000000000000001010',
     *       '',
     *       '0xe0b7927c4af23765cb51314a0e0521a9645f0e2a' // DGD contract address
     * );
     * @returns {Promise.<object>}
     */
    tokenbalance(address, tokenname, contractaddress) {

      const module = 'account';
      const action = 'tokenbalance';
      const tag = 'latest';

      var queryObject = {
        module, action, apiKey, tag
      };

      if (contractaddress) {
        queryObject.contractaddress = contractaddress;
      }

      if (tokenname) {
        queryObject.tokenname = tokenname;
      }

      if (address) {
        queryObject.address = address;
      }

      var query = querystring.stringify(queryObject);
      return getRequest(query);
    },
    /**
     * Returns the balance of a sepcific account
     * @param {string} address - Address
     * @example
     * var balance = api.account.balance('0x0000000000000000000000000000000000001010');
     * @returns {Promise.<object>}
     */
    balance(address) {
      const module = 'account';
      let action = 'balance';
      const tag = 'latest';

      if (typeof address !== 'string' && address && address.length) {
        address = address.join(',');
        action = 'balancemulti';
      }

      var query = querystring.stringify({
        module, action, tag, address, apiKey
      });
      return getRequest(query);
    },
    /**
     * Get a list of internal transactions
     * @param {string} txhash - Transaction hash. If specified then address will be ignored
     * @param {string} address - Transaction address
     * @param {string} startblock - start looking here
     * @param {string} endblock - end looking there
     * @param {string} sort - Sort asc/desc
     * @example
     * var txlist = api.account.txlistinternal('0x40eb908387324f2b575b4879cd9d7188f69c8fc9d87c901b9e2daaea4b442170');
     * @returns {Promise.<object>}
     */
    txlistinternal(txhash, address, startblock, endblock, sort) {
      const module = 'account';
      const action = 'txlistinternal';

      var queryObject = {
        module,
        action,
        apiKey
      };

      if (!startblock) {
        startblock = 0;
      }
      queryObject.startblock = startblock;

      if (!endblock) {
        endblock = 'latest';
      }
      queryObject.endblock = endblock;

      if (!sort) {
        sort = 'asc';
      }
      queryObject.sort = sort;

      if (txhash) {
        queryObject.txhash = txhash;
      } else {
        queryObject.address = address;
      }

      queryObject.txhash = txhash;

      return getRequest(querystring.stringify(queryObject));
    },
    /**
     * Get a list of transactions for a specfic address
     * @param {string} address - Transaction address
     * @param {string} startblock - start looking here
     * @param {string} endblock - end looking there
     * @param {number} page - Page number
     * @param {number} offset - Max records to return
     * @param {string} sort - Sort asc/desc
     * @example
     * var txlist = api.account.txlist('0x0000000000000000000000000000000000001010', 1, 'latest', 1, 100, 'asc');
     * @returns {Promise.<object>}
     */
    txlist(address, startblock, endblock, page, offset, sort) {
      const module = 'account';
      const action = 'txlist';

      if (!startblock) {
        startblock = 0;
      }

      if (!endblock) {
        endblock = 'latest';
      }

      if (!page) {
        page = 1;
      }

      if (!offset) {
        offset = 100;
      }

      if (!sort) {
        sort = 'asc';
      }

      var query = querystring.stringify({
        module, action, startblock, endblock, page, offset, sort, address, apiKey
      });

      return getRequest(query);
    },
    /**
     * Get a list of blocks that a specific account has mineds
     * @example
     * var txlist = api.account.getminedblocks('0x46a3a41bd932244dd08186e4c19f1a7e48cbcdf4');
     * @param {string} address - Transaction hash
     */
    getminedblocks(address) {
      const module = 'account';
      const action = 'getminedblocks';
      var query = querystring.stringify({
        module, action, address, apiKey
      });
      return getRequest(query);
    },
     /**
     * [BETA] Get a list of "ERC20 - Token Transfer Events" by Address
     * @param {string} address - Account address
     * @param {string} startblock - start looking here
     * @param {string} endblock - end looking there
      * @param {number} page - Page number
      * @param {number} offset - Max records to return
      * @param {string} sort - Sort asc/desc
     * @param {string} contractaddress - Address of ERC20 token contract (if not specified lists transfers for all tokens)
     * @example
     * var txlist = api.account.tokentx('0x0000000000000000000000000000000000001010', '0x2791bca1f2de4661ed88a30c99a7a9449aa84174', 1, 'latest', 'asc');
     * @returns {Promise.<object>}
     */
    tokentx(address, contractaddress, startblock, endblock, page, offset, sort) {
      const module = 'account';
      const action = 'tokentx';

      if (!startblock) {
        startblock = 0;
      }

      if (!endblock) {
        endblock = 'latest';
      }

       if (!page) {
         page = 1;
       }

       if (!offset) {
         offset = 100;
       }

      if (!sort) {
        sort = 'asc';
      }

      var queryObject = {
        module, action, startblock, endblock, page, offset, sort, address, apiKey
      };

      if (contractaddress) {
        queryObject.contractaddress = contractaddress;
      }

      return getRequest(querystring.stringify(queryObject));
    }
  };
};
